1.18.2 Manual Compressor

Alice0Rose opened the pull request:
Adds the Manual Compressor discussed in https://github.com/TeamPneumatic/pnc-repressurized/issues/1112.
- Generates compressed air from manual labor (right clicking).
- Generates a static amount of air each pump cycle, but the more air in the compressor, the more right clicks are needed to complete each pump cycle.
- Impossible to generate air at pressure greater than 5 bar.
- [Preview video](https://youtu.be/vJUMcFR-6RU)

Commit:
Adds Manual Compressor

TeamSpen210 commented:
It'd be good to make this drain your hunger with each click - I know a lot of other mods do that for manual devices, gives it a cost as well as makes sense since the player would be putting in the effort.

desht commented:
Agreed - the amount of hunger could also be a configurable for pack makers to tweak as needed.

desht commented on src/main/java/me/desht/pneumaticcraft/common/block/entity/ManualCompressorBlockEntity.java:
```diff
@@ -0,0 +1,94 @@
+    @Override
+    public void tickServer() {
+        super.tickServer();
+        storedAir = airHandler.getAir();
```
What precision does the client actually need here?  Sync'ing this every time it changes could lead to a lot of chatter.  Clamping the sync'd value to the nearest X will reduce traffic (and thus lag).

desht commented on src/main/java/me/desht/pneumaticcraft/common/block/entity/ManualCompressorBlockEntity.java:
```diff
@@ -0,0 +1,94 @@
+     * Triggered on completion of a pump
+     * Adds compressed air from pump and resets pump cycle
+     */
+    public void onPumpCycleComplete() {
```
As mentioned in the PR comments, a configurable player hunger here would be nice to have. Probably reasonably low by default.

TeamSpen210 replied:
It might be better to do it in `onPumpCycle()`, so  it's more of a continuous drain and naturally becomes more expensive as the pressure increases?

desht replied:
Yeah, sounds better.

desht commented on src/main/java/me/desht/pneumaticcraft/common/block/ManualCompressorBlock.java:
```diff
@@ -0,0 +1,90 @@
+        BlockEntity be = world.getBlockEntity(pos);
+
+        // Triggers a pump cycle when manual compressor is right-clicked
+        if (be instanceof ManualCompressorBlockEntity manualCompressorBlockEntity) {
```
We probably want to disallow fake players here?  Otherwise this becomes a trivial source of pressure, depending on what other mods are available.

This could be a configurable (default to disallow fake players).

desht commented on src/main/resources/assets/pneumaticcraft/patchouli_books/book/en_us/entries/compressors/maunal_compressor.json:
```diff
@@ -0,0 +1,17 @@
+  "advancement": "pneumaticcraft:pressure_tube",
+  "pages": [
+    {
+      "type": "text",
```
Hunger requirements, once added, should be documented here.

desht commented on src/main/java/me/desht/pneumaticcraft/client/render/blockentity/ManualCompressorRenderer.java:
```diff
@@ -0,0 +1,85 @@
+
+        // Changes the position of the pumprod relative to the pump cycle's progress
+        // 0% = 0, 100% = 6.75
+        double pumprodVerticalOffset = (te.pumpCycleProgress / (400/27.0));
```
I think it should be possible to use `partialTicks` here to get a smoother pump animation.  I noticed from your video that it's quite jerky, particularly early on when it's quick.

Alice0Rose commented:
> It'd be good to make this drain your hunger with each click - I know a lot of other mods do that for manual devices, gives it a cost as well as makes sense since the player would be putting in the effort.

Sounds like a good idea, and a good way of further balancing something that is meant to be one of the most inefficient compressors.

> Agreed - the amount of hunger could also be a configurable for pack makers to tweak as needed.

I'll also make the amount of air produced each cycle configurable. I meant to originally but I think I just forgot to.

Commit:
- Made pump animation smoother for manual compressor
- Made manual compressor consume hunger for each pump step
- Added configs for manual compressor to control how much air is produced, how much hunger is used, and if fake players can operate it
- Limited speed at which manual compressor can be used to prevent auto-clicker abuse
- Made storedAir sync less frequently to client
- Added hunger requirements to guide book and manual compressor description

Alice0Rose commented:
[Preview video](https://youtu.be/KLNN2fA9_gY) showing off the updated manual compressor

desht commented:
Merged, thanks!
